### PR TITLE
feat : wishlist screen with functionality

### PR DESCRIPTION
**Pull request by vanshitalavange**
Overview: 
- wishlist screen
- wishlist functionality
- refactored auth 

Preview:
https://mandeshi-agro-git-wishlist-vanshitalavange.vercel.app/

### CODE REVIEW

File: src/components/Header/Header.jsx [hunk 1]
```diff
@@ -4,10 +4,16 @@ import { Link, useNavigate } from "react-router-dom";
...
 export function Header() {
+ 
+  const { userState, setUserState } = useAuth();
+  const { loginStatus,  userDetails } = userState;
+  const { wishlist } = userDetails;
   const navigate = useNavigate();
+
   const { setShowResponsiveNavbarForMobile } = useNavbar();
-  const { isLoggedIn, setIsLoggedIn, userDetails } = useAuth();
+
```
Comment (ttahm3d) [inline src/components/Header/Header.jsx:14]: In mobile view, your logo is getting shown in center. Kindly check this. Problem is with `justify: space-between`. Use margin left: auto on the right side cart/wishlist and other badges

File: src/components/Header/Header.jsx [hunk 2]
```diff
@@ -93,15 +96,15 @@ export function Header() {
               <div className="badge">
                 <i className="fa fa-heart header-icon header-badge-icon"></i>
                 <span className="badge-counter badge-round badge-counter-right">
-                  0
+                  {loginStatus ? wishlist.length : 0}
```
Comment (ttahm3d) [inline src/components/Header/Header.jsx]: It doesn't make sense to display 0 if the user is not logged in. Use null instead.

File: src/components/ProductCard/ProductCard.jsx [hunk 3]
```diff
@@ -1,23 +1,73 @@
...
+
+  const { _id, productName, price, prevPrice, imgSrc, rating } = { ...product };
+
+  const [isWishlisted, setIsWishlisted] = useState({
+    value: false,
+    class: "not-wishlisted",
+  });
+
+  const isProductInWishlist = (wishlist, product) => {
+    const result = wishlist.filter((item) => item._id === product._id);
```
Comment (ttahm3d) [inline src/components/ProductCard/ProductCard.jsx]: Read about `some` funciton. It makes this functionality easier.

File: src/components/ProductCard/ProductCard.jsx [hunk 4]
```diff
@@ -1,23 +1,73 @@
...
+      const removeFromWishlistAPIResponse = removeFromWishlist(authToken, _id);
+      removeFromWishlistAPIResponse.then((data) =>
+        setUserState({
+          ...userState,
+          userDetails: { ...userDetails, wishlist: data },
+        })
+      );
+      setIsWishlisted({ value: false, class: "not-wishlisted" });
+    }
+  };
```
Comment (ttahm3d) [inline src/components/ProductCard/ProductCard.jsx]: It would be better to create these inside context and pass it down via value of context

**Review by ttahm3d — COMMENTED** [review]
Looks good. Have provided some comments. Please take a look.

File: src/components/Header/Header.jsx (hunk 1 above)
Comment (vanshitalavange) [reply src/components/Header/Header.jsx:14]: I wanted my logo to be in the center in mobile view

File: src/components/Header/Header.jsx (hunk 2 above)
Comment (vanshitalavange) [reply src/components/Header/Header.jsx]: changes done

File: src/components/ProductCard/ProductCard.jsx (hunk 3 above)
Comment (vanshitalavange) [reply src/components/ProductCard/ProductCard.jsx]: yes it did reduce my work, changes done

File: src/components/ProductCard/ProductCard.jsx (hunk 4 above)
Comment (vanshitalavange) [reply src/components/ProductCard/ProductCard.jsx]: done

**Review by dev-enforced — APPROVED** [review]
Liked the way you exported the utilities related to authentication refactoring and creating separate reducers related to wishlist!!
Got a new approach on how to implement Protected Routes !!
@ttahm3d suggestions also cleared out certain things which you have worked out!!
The functionality is working as expected on this site: https://mandeshi-agro-git-wishlist-vanshitalavange.vercel.app/wishlist

**Pull request merged**